api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+const (
+	Memory = iota
+	CPU
+)
+
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		// deployCache: map[string][]*Resource,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	humanReadableProviders := ""
+	for _, r := range resourceProviders {
+		humanReadableProviders += fmt.Sprintf("\n    %#v", r)
+	}
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceProviders for env '%s': %s", environmentID, humanReadableProviders)
+
+	var resourceConsumers []ResourceConsumer
+
+	// get pending service resource consumers
+	serviceConsumers, err := e.getResourceConsumers_PendingServices(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, serviceConsumers...)
+
+	// get pending task resource consumers in ECS
+	ecsTaskConsumers, err := e.getResourceConsumers_TasksInECS(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, ecsTaskConsumers...)
+
+	// get pending task resource consumers in jobs
+	jobTaskConsumers, err := e.getResourceConsumers_TasksInJobs(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, jobTaskConsumers...)
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	var errs []error
+
+	// calculate for scaling up
+	resourceProviders, scaleUpErrs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+
+	errs = append(errs, scaleUpErrs...)
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+
+	// do the scaling
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	if err := e.scaleTo(clusterName, desiredScale, unusedProviders); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.MultiError(errs)
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*ResourceProvider, error) {
+	env, err := e.EnvironmentProvider.Read(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &ResourceProvider{}
+	resource.AvailableMemory = InstanceSizes[env.InstanceSize]
+	resource.ID = "<new instance>"
+	resource.InUse = false
```

Comment:
True - if this makes things a little easier to understand/follow, I'd probably opt to keep it, but otherwise I have no problem stripping it out. @sesh-kebab

Reply:
This resource creation has been refactored to use a constructor from `scaler/resource_provider.go`. ✅